django-app/donations/payment_gateways/paypal/views.py

```diff
@@ -221,3 +221,87 @@ def cancel_from_paypal(request):
         request.session['error-message'] = str(error)
         _exception(str(error))
     return redirect('donations:cancelled')
+
+
+def return_from_paypal_card_update(request):
+    """ This endpoint is submitted as the return_url in the updating the payment flow.
+        This url should receive GET params: 'token' and 'subscription_id'(only recurring payments); 'ba_token' is not used
+        In Factory_PayPal.initGatewayByReturn(request), we save the subscription_id/token into the donationPaymentMeta data upon a successful request;
+        exception will be raised if the endpoint is reached but a previous meta value is found,
+        this is done to prevent this endpoint being called unlimitedly
+
+        @todo: revise error handling, avoid catching all exceptions at the end
+    """
+    try:
+        gatewayManager = Factory_Paypal.initGatewayByReturn(request)
+        if gatewayManager.donation.is_recurring:
+            gatewayManager.donation.subscription.profile_id = request.GET.get('subscription_id')
+            gatewayManager.donation.subscription.save()
+            try:
+                # We want to cancel the payment failed subscriptioninstance under this subscription
+                # now that we have an active one
+                parent = gatewayManager.donation.subscription.parent
+                instance = SubscriptionInstance.objects.get(parent=parent, recurring_status=STATUS_PAYMENT_FAILED)
+
+                # Save old_instance_id with the subscription id so that we don't send emails
+                spmeta = SubscriptionPaymentMeta(
+                    subscription=instance, field_key='old_instance_id', field_value=instance.id)
```

Comment:
This `SubscriptionPaymentMeta` record created here is actually different from the one we created in donation.views, I was confused at first because the same `field_key` "old_instance_id" is used, but on closer inspection the `subscription` points to a different instance.

Let's rename this key here as something else (e.g. "awaiting_cancellation") to avoid confusion?